pkg/buildexpression/parser/parser.go

```diff
@@ -0,0 +1,541 @@
+	}
+
+	for p.tok != EOF {
+		switch p.tok {
+		case LET:
+			err = p.parseExpression(result.Root)
+			if err != nil {
+				return nil, errs.Wrap(err, "Failed to parse expression")
+			}
+			if p.tok == R_CURL {
+				result.Root.AddChild(p.newNode(NodeRightCurlyBracket))
+			}
+		case IN:
+			err = p.parseIn(result.Root)
+			if err != nil {
+				return nil, errs.Wrap(err, "Failed to parse in")
+			}
+			if p.tok == COMMA {
+				result.Root.AddChild(p.newNode(NodeComma))
+			}
+		}
+
+		p.next()
+	}
+
+	return &result, nil
+}
+
+func (p *Parser) parseExpression(root *Node) error {
+	if !isExpression(p.tok) {
+		return errs.New("Expected expression")
+	}
+	expressionNode := p.newNode(NodeExpression)
+	root.AddChild(expressionNode)
+
+	expressionNode.AddChild(p.newNode(NodeLet))
+
+	p.next()
+
+	err := p.expectToken(COLON, expressionNode, NodeColon)
+	if err != nil {
+		return errs.Wrap(err, "Expect failed")
+	}
+
+	err = p.expectToken(L_CURL, expressionNode, NodeLeftCurlyBracket)
+	if err != nil {
+		return errs.Wrap(err, "Expect failed")
+	}
+
+	err = p.parseBinding(expressionNode)
+	if err != nil {
+		return errs.Wrap(err, "Failed to parse binding")
+	}
+
+	return nil
+}
+
+func (p *Parser) parseBinding(node *Node) error {
+	bindingNode := p.newNode(NodeBinding)
+	node.AddChild(bindingNode)
+
+	for {
+		if !p.isAssignment() {
+			break
+		}
+
+		err := p.parseAssignment(bindingNode)
+		if err != nil {
+			return errs.Wrap(err, "Failed to parse assignment")
+		}
+	}
+
+	return nil
+}
+
+func (p *Parser) parseAssignment(node *Node) error {
+	if !p.isAssignment() {
+		return errs.New("Expected assignment")
+	}
+
+	assignmentNode := p.newNode(NodeAssignment)
+	node.AddChild(assignmentNode)
+
+	err := p.parseIdentifier(assignmentNode)
+	if err != nil {
+		return errs.Wrap(err, "Failed to parse identifier")
+	}
+
+	err = p.expectToken(COLON, assignmentNode, NodeColon)
+	if err != nil {
+		return errs.Wrap(err, "Expect failed")
+	}
+
+	// In this case the next token should be a function identifier
+	if p.tok == L_CURL && isFunctionIdentifier(p.peek()) {
+		assignmentNode.AddChild(p.newNode(NodeLeftCurlyBracket))
+		p.next()
+
+		if !isFunctionIdentifier(p.tok) {
+			return errs.New("Expected function identifier")
+		}
+
+		err = p.parseApplication(assignmentNode)
+		if err != nil {
+			return errs.Wrap(err, "Failed to parse application")
+		}
+
+		err = p.expectToken(R_CURL, assignmentNode, NodeRightCurlyBracket)
+		if err != nil {
+			return errs.Wrap(err, "Expect failed")
+		}
+
+		if p.tok == COMMA {
+			assignmentNode.AddChild(p.newNode(NodeComma))
+			p.next()
+		}
+	}
+
+	var msg string
+	switch p.tok {
+	case STRING:
+		err = p.parseString(assignmentNode)
+		msg = "Failed to parse string"
+	case IDENTIFIER:
+		err = p.parseIdentifier(assignmentNode)
+		msg = "Failed to parse identifier"
+	case IN, R_CURL:
+		// If we encounter an IN or R_CURL token then we've reached the end of the binding
+		return nil
+	default:
```

Comment:
Would you consider having an explicit case for lists here, and defaulting to an error for unexpected input?